File: src/main/java/net/kyori/text/Component.java
/*
 * This file is part of text, licensed under the MIT License.
 *
 * Copyright (c) 2017-2018 KyoriPowered
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package net.kyori.text;

import net.kyori.blizzard.NonNull;
import net.kyori.blizzard.Nullable;
import net.kyori.text.event.ClickEvent;
import net.kyori.text.event.HoverEvent;
import net.kyori.text.format.TextColor;
import net.kyori.text.format.TextDecoration;

import java.util.Collections;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

/**
 * A text component.
 */
public interface Component {
  /**
   * An empty, unmodifiable, list of components.
   */
  List<Component> EMPTY_COMPONENT_LIST = Collections.emptyList();

  /**
   * Gets the unmodifiable list of children.
   *
   * @return the unmodifiable list of children
   */
  @NonNull
  List<Component> children();

  /**
   * Checks if this component contains a component.
   *
   * @param that the other component
   * @return {@code true} if this component contains the provided
   *     component, {@code false} otherwise
   */
  default boolean contains(@NonNull final Component that) {
    if(this == that) return true;
    for(final Component child : this.children()) {
      if(child.contains(that)) return true;
    }
    if(this.hoverEvent() != null) {
      final Component hover = this.hoverEvent().value();
      if(that == hover) return true;
      for(final Component child : hover.children()) {
        if(child.contains(that)) return true;
      }
    }
    return false;
  }

  /**
   * Prevents a cycle between this component and the provided component.
   *
   * @param that the other component
   */
  default void detectCycle(@NonNull final Component that) {
    if(that.contains(this)) {
      throw new IllegalStateException("Component cycle detected between " + this + " and " + that);
    }
  }

  /**
   * Appends a component to this component.
   *
   * @param component the component to append
   * @return a component with the component added
   */
  @NonNull
  Component append(@NonNull final Component component);

  /**
   * Creates a component.
   *
   * @return a component
   */
  @NonNull
  Component copy();

  /**
   * Gets the color of this component.
   *
   * @return the color of this component
   */
  @Nullable
  TextColor color();

  /**
   * Sets the color of this component.
   *
   * @param color the color
   * @return a component
   */
  @NonNull
  Component color(@Nullable final TextColor color);

  /**
   * Tests if this component has a decoration.
   *
   * @param decoration the decoration
   * @return {@code true} if this component has the decoration, {@code false} if this
   *     component does not have the decoration
   */
  default boolean hasDecoration(@NonNull final TextDecoration decoration) {
    return this.decoration(decoration) == TextDecoration.State.TRUE;
  }

  /**
   * Gets the state of a decoration on this component.
   *
   * @param decoration the decoration
   * @return {@link TextDecoration.State#TRUE} if this component has the decoration,
   *     {@link TextDecoration.State#FALSE} if this component does not have the decoration,
   *     and {@link TextDecoration.State#NOT_SET} if not set
   */
  @NonNull
  TextDecoration.State decoration(@NonNull final TextDecoration decoration);

  /**
   * Sets the state of a decoration on this component.
   *
   * @param decoration the decoration
   * @param flag {@code true} if this component should have the decoration, {@code false} if
   *     this component should not have the decoration
   * @return a component
   */
  @NonNull
  default Component decoration(@NonNull final TextDecoration decoration, final boolean flag) {
    return this.decoration(decoration, TextDecoration.State.byBoolean(flag));
  }

  /**
   * Sets the value of a decoration on this component.
   *
   * @param decoration the decoration
   * @param state {@link TextDecoration.State#TRUE} if this component should have the
   *     decoration, {@link TextDecoration.State#FALSE} if this component should not
   *     have the decoration, and {@link TextDecoration.State#NOT_SET} if the decoration
   *     should not have a set value
   * @return a component
   */
  @NonNull
  Component decoration(@NonNull final TextDecoration decoration, @NonNull final TextDecoration.State state);

  /**
   * Gets a set of decorations this component has.
   *
   * @return a set of decorations this component has
   */
  @NonNull
  default Set<TextDecoration> decorations() {
    return this.decorations(Collections.emptySet());
  }

  /**
   * Gets a set of decorations this component has.
   *
   * @param defaultValues a set of default values
   * @return a set of decorations this component has
   */
  @NonNull
  default Set<TextDecoration> decorations(@NonNull final Set<TextDecoration> defaultValues) {
    final Set<TextDecoration> decorations = EnumSet.noneOf(TextDecoration.class);
    for(final TextDecoration decoration : TextDecoration.values()) {
      final TextDecoration.State value = this.decoration(decoration);
      if(value == TextDecoration.State.TRUE || (value == TextDecoration.State.NOT_SET && defaultValues.contains(decoration))) {
        decorations.add(decoration);
      }
    }
    return decorations;
  }

  /**
   * Gets the click event of this component.
   *
   * @return the click event
   */
  @Nullable
  ClickEvent clickEvent();

  /**
   * Sets the click event of this component.
   *
   * @param event the click event
   * @return a component
   */
  @NonNull
  Component clickEvent(@Nullable final ClickEvent event);

  /**
   * Gets the hover event of this component.
   *
   * @return the hover event
   */
  @Nullable
  HoverEvent hoverEvent();

  /**
   * Sets the hover event of this component.
   *
   * @param event the hover event
   * @return a component
   */
  @NonNull
  Component hoverEvent(@Nullable final HoverEvent event);

  /**
   * Gets the string to be inserted when this component is shift-clicked.
   *
   * @return the insertion string
   */
  @Nullable
  String insertion();

  /**
   * Sets the string to be inserted when this component is shift-clicked.
   *
   * @param insertion the insertion string
   * @return a component
   */
  @NonNull
  Component insertion(@Nullable final String insertion);

  /**
   * Merges styling from another component into this component.
   *
   * @param that the other component
   * @return a component
   */
  @NonNull
  Component mergeStyle(@NonNull final Component that);

  /**
   * Merges the color from another component into this component.
   *
   * @param that the other component
   * @return a component
   */
  @NonNull
  Component mergeColor(@NonNull final Component that);

  /**
   * Merges the decorations from another component into this component.
   *
   * @param that the other component
   * @return a component
   */
  @NonNull
  Component mergeDecorations(@NonNull final Component that);

  /**
   * Merges the events from another component into this component.
   *
   * @param that the other component
   * @return a component
   */
  @NonNull
  Component mergeEvents(@NonNull final Component that);

  /**
   * Resets all styling on this component.
   *
   * @return a component
   */
  @NonNull
  Component resetStyle();

  /**
   * Tests if this component has any styling.
   *
   * @return {@code true} if this component has any styling, {@code false} if this
   *     component does not have any styling
   */
  boolean hasStyling();

  /**
   * Creates a text component with the content of {@link String#valueOf(boolean)}.
   *
   * @param value the boolean value
   * @return the component
   */
  @NonNull
  static Component of(final boolean value) {
    return TextComponent.of(String.valueOf(value));
  }

  /**
   * Creates a text component with the content of {@link String#valueOf(char)}.
   *
   * @param value the char value
   * @return the component
   */
  @NonNull
  static Component of(final char value) {
    return TextComponent.of(String.valueOf(value));
  }

  /**
   * Creates a text component with the content of {@link String#valueOf(double)}.
   *
   * @param value the double value
   * @return the component
   */
  @NonNull
  static Component of(final double value) {
    return TextComponent.of(String.valueOf(value));
  }

  /**
   * Creates a text component with the content of {@link String#valueOf(float)}.
   *
   * @param value the float value
   * @return the component
   */
  @NonNull
  static Component of(final float value) {
    return TextComponent.of(String.valueOf(value));
  }

  /**
   * Creates a text component with the content of {@link String#valueOf(int)}.
   *
   * @param value the int value
   * @return the component
   */
  @NonNull
  static Component of(final int value) {
    return TextComponent.of(String.valueOf(value));
  }

  /**
   * Creates a text component with the content of {@link String#valueOf(long)}.
   *
   * @param value the long value
   * @return the component
   */
  @NonNull
  static Component of(final long value) {
    return TextComponent.of(String.valueOf(value));
  }
}

File: api-kotlin/src/main/kotlin/net/kyori/adventure/extension/kotlin/Component.kt
/*
 * This file is part of adventure, licensed under the MIT License.
 *
 * Copyright (c) 2017-2020 KyoriPowered
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package net.kyori.adventure.extension.kotlin

import net.kyori.adventure.key.Key
import net.kyori.adventure.text.BlockNBTComponent
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.ComponentBuilder
import net.kyori.adventure.text.ComponentLike
import net.kyori.adventure.text.EntityNBTComponent
import net.kyori.adventure.text.KeybindComponent
import net.kyori.adventure.text.ScoreComponent
import net.kyori.adventure.text.SelectorComponent
import net.kyori.adventure.text.StorageNBTComponent
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.TextComponent.empty
import net.kyori.adventure.text.TranslatableComponent
import net.kyori.adventure.text.format.Style
import net.kyori.adventure.text.format.StyleBuilderApplicable

// Joining components together //

// elipses for truncated joins
internal val TRUNCATE_MARK = TextComponent.of("...")
internal val COMMA_SPACE = TextComponent.of(", ")

/**
 * Add [that] as a child component.
 */
public operator fun Component.plus(that: ComponentLike): Component = this.append(that)

/**
 * Append [that] as a child of the component being built.
 */
public operator fun ComponentBuilder<*, *>.plusAssign(that: ComponentLike) {
  this.append(that)
}

/**
 * Append all components in [that] as children of the component being built.
 */
public operator fun ComponentBuilder<*, *>.plusAssign(that: Iterable<ComponentLike>) {
  this.append(that)
}

/**
 * Convert this object into a component
 */
public operator fun ComponentLike.unaryPlus(): Component = asComponent()

/**
 * Append the [Iterable] of components to an existing builder, returning the built component
 *
 * This method should match the specification of [kotlin.collections.joinTo], but
 * acting on [Component]s rather than on Strings.
 */
public fun <T: ComponentLike, B: ComponentBuilder<*, B>> Iterable<T>.joinTo(
  builder: B,
  joiner: Component = COMMA_SPACE,
  prefix: Component = empty(),
  suffix: Component = empty(),
  transform: (Component) -> Component = { it },
  limit: Int = -1,
  truncateWith: Component = TRUNCATE_MARK
): B {
  val iter = iterator()
  builder.append(prefix)

  var count = 0
  while(iter.hasNext()) {
    if(limit <= 0 || count++ < limit) {
      builder.append(transform(iter.next().asComponent()))
    } else {
      builder.append(truncateWith)
      break
    }

    if(iter.hasNext()) {
      builder.append(joiner)
    }
  }

  builder.append(suffix)
  return builder
}

/**
 * Join an iterable of components into a new [TextComponent]
 *
 * @see [joinTo] for parameter descriptions
 */
public fun <T: ComponentLike> Iterable<T>.join(
  joiner: Component = COMMA_SPACE,
  prefix: Component = empty(),
  suffix: Component = empty(),
  transform: (Component) -> Component = { it },
  limit: Int = -1,
  truncateWith: Component = TRUNCATE_MARK
): Component = joinTo(TextComponent.builder(), joiner, prefix, suffix, transform, limit, truncateWith).build()

// Factory methods //

/** Create a new text component from `this` */
public fun String.text(vararg styles: StyleBuilderApplicable): TextComponent = TextComponent.of(this, Style.of(*styles))

/** Create a new translatable component from `this` */
public fun String.tr(vararg args: ComponentLike): TranslatableComponent = TranslatableComponent.of(this, *args)

/** Create a new keybind component using the key sequence from `this` */
public fun String.keybind(vararg styles: StyleBuilderApplicable): KeybindComponent = KeybindComponent.of(this, Style.of(*styles))

/** Create a new selector component, using `this` as a selector. */
public fun String.selector(vararg styles: StyleBuilderApplicable): SelectorComponent = SelectorComponent.builder(this).style(Style.of(*styles)).build()

/** Create a new score component, with the score at `this` from [objective] */
public fun String.score(objective: String, vararg styles: StyleBuilderApplicable): ScoreComponent = ScoreComponent.builder(this, objective).style(Style.of(*styles)).build()

/** Create a new block NBT component, with nbt path from `this` gotten at [pos] */
public fun String.blockNBT(pos: BlockNBTComponent.Pos, interpret: Boolean = false, vararg styles: StyleBuilderApplicable): BlockNBTComponent = BlockNBTComponent.builder()
  .nbtPath(this)
  .pos(pos)
  .interpret(interpret)
  .style(Style.of(*styles))
  .build()

/** Create a new entity NBT component, with nbt path from `this` gotten from the entity marked by [entitySelector] */
public fun String.entityNBT(entitySelector: String, interpret: Boolean = false, vararg styles: StyleBuilderApplicable): EntityNBTComponent = EntityNBTComponent.builder()
  .nbtPath(this)
  .selector(entitySelector)
  .interpret(interpret)
  .style(Style.of(*styles))
  .build()

/** Create a new storage NBT component, with nbt path from `this` gotten from the named storage at [storage] */
public fun String.storageNBT(storage: Key, interpret: Boolean = false, vararg styles: StyleBuilderApplicable): StorageNBTComponent = StorageNBTComponent.builder()
  .nbtPath(this)
  .storage(storage)
  .interpret(interpret)
  .style(Style.of(*styles))
  .build()
